File: frontend/apps/ios/Pantopus/Core/Networking/Models/Settings/SettingsDTOs.swift
//
//  SettingsDTOs.swift
//  Pantopus
//
//  Decoder shapes for the privacy + auth-methods Settings endpoints.
//  All fields are Optional so the backend can roll out new keys
//  without breaking older clients.
//

import Foundation

/// Envelope from `GET /api/privacy/settings`.
public struct PrivacySettingsResponse: Decodable, Sendable {
    public let settings: PrivacySettings
}

/// One privacy-settings row.
public struct PrivacySettings: Decodable, Sendable, Hashable {
    public let userId: String?
    public let searchVisibility: String?
    public let addressPrecision: String?
    public let hideFromSearch: Bool?
    public let showOnlineStatus: Bool?
    public let showLastActive: Bool?
    public let showReadReceipts: Bool?
    public let shareHomeCheckIns: Bool?
    public let pushPreferences: [String: Bool]?
    public let emailPreferences: [String: Bool]?
    public let smsPreferences: [String: Bool]?
    public let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case searchVisibility = "search_visibility"
        case addressPrecision = "address_precision"
        case hideFromSearch = "hide_from_search"
        case showOnlineStatus = "show_online_status"
        case showLastActive = "show_last_active"
        case showReadReceipts = "show_read_receipts"
        case shareHomeCheckIns = "share_home_check_ins"
        case pushPreferences = "push_preferences"
        case emailPreferences = "email_preferences"
        case smsPreferences = "sms_preferences"
        case updatedAt = "updated_at"
    }
}

/// Envelope from `GET /api/privacy/blocks`.
public struct PrivacyBlocksResponse: Decodable, Sendable {
    public let blocks: [PrivacyBlock]
}

public struct PrivacyBlock: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let blockedUserId: String?
    public let blockScope: String?
    public let reason: String?
    public let createdAt: String?
    public let blocked: BlockedUserSummary?

    enum CodingKeys: String, CodingKey {
        case id
        case blockedUserId = "blocked_user_id"
        case blockScope = "block_scope"
        case reason
        case createdAt = "created_at"
        case blocked
    }
}

/// Nested user summary returned by `privacy.js:154` join.
public struct BlockedUserSummary: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let username: String?
    public let name: String?
    public let profilePictureUrl: String?

    enum CodingKeys: String, CodingKey {
        case id, username, name
        case profilePictureUrl = "profile_picture_url"
    }
}

/// Envelope from `GET /api/users/auth-methods`.
public struct AuthMethodsResponse: Decodable, Sendable {
    public let methods: [AuthMethod]?
    public let hasPassword: Bool?
    public let providers: [String]?
    public let twoFactorEnabled: Bool?

    enum CodingKeys: String, CodingKey {
        case methods
        case hasPassword = "has_password"
        case providers
        case twoFactorEnabled = "two_factor_enabled"
    }
}

public struct AuthMethod: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let provider: String?
    public let label: String?

    enum CodingKeys: String, CodingKey {
        case id, provider, label
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/settings/SettingsDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.settings

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/** Envelope from `GET /api/privacy/settings`. */
@JsonClass(generateAdapter = true)
data class PrivacySettingsResponse(
    val settings: PrivacySettingsDto,
)

/** One privacy-settings row. All fields are nullable so the backend
 * can roll out new keys without breaking older clients. */
@JsonClass(generateAdapter = true)
data class PrivacySettingsDto(
    @Json(name = "user_id") val userId: String? = null,
    @Json(name = "search_visibility") val searchVisibility: String? = null,
    @Json(name = "address_precision") val addressPrecision: String? = null,
    @Json(name = "hide_from_search") val hideFromSearch: Boolean? = null,
    @Json(name = "show_online_status") val showOnlineStatus: Boolean? = null,
    @Json(name = "show_last_active") val showLastActive: Boolean? = null,
    @Json(name = "show_read_receipts") val showReadReceipts: Boolean? = null,
    @Json(name = "share_home_check_ins") val shareHomeCheckIns: Boolean? = null,
    @Json(name = "push_preferences") val pushPreferences: Map<String, Boolean>? = null,
    @Json(name = "email_preferences") val emailPreferences: Map<String, Boolean>? = null,
    @Json(name = "sms_preferences") val smsPreferences: Map<String, Boolean>? = null,
    @Json(name = "updated_at") val updatedAt: String? = null,
)

/** Partial-update body for `PATCH /api/privacy/settings`. */
@JsonClass(generateAdapter = true)
data class PrivacySettingsUpdate(
    @Json(name = "search_visibility") val searchVisibility: String? = null,
    @Json(name = "address_precision") val addressPrecision: String? = null,
    @Json(name = "hide_from_search") val hideFromSearch: Boolean? = null,
    @Json(name = "show_online_status") val showOnlineStatus: Boolean? = null,
    @Json(name = "show_last_active") val showLastActive: Boolean? = null,
    @Json(name = "show_read_receipts") val showReadReceipts: Boolean? = null,
    @Json(name = "share_home_check_ins") val shareHomeCheckIns: Boolean? = null,
    @Json(name = "push_preferences") val pushPreferences: Map<String, Boolean>? = null,
    @Json(name = "email_preferences") val emailPreferences: Map<String, Boolean>? = null,
    @Json(name = "sms_preferences") val smsPreferences: Map<String, Boolean>? = null,
)

/** Envelope from `GET /api/privacy/blocks`. */
@JsonClass(generateAdapter = true)
data class PrivacyBlocksResponse(
    val blocks: List<PrivacyBlockDto>,
)

@JsonClass(generateAdapter = true)
data class PrivacyBlockDto(
    val id: String,
    @Json(name = "blocked_user_id") val blockedUserId: String? = null,
    @Json(name = "block_scope") val blockScope: String? = null,
    val reason: String? = null,
    @Json(name = "created_at") val createdAt: String? = null,
    val blocked: BlockedUserSummaryDto? = null,
)

/** Nested user summary returned by `privacy.js:154` join. */
@JsonClass(generateAdapter = true)
data class BlockedUserSummaryDto(
    val id: String,
    val username: String? = null,
    val name: String? = null,
    @Json(name = "profile_picture_url") val profilePictureUrl: String? = null,
)

/** Body for `POST /api/notifications/push-token`. */
@JsonClass(generateAdapter = true)
data class PushTokenBody(
    val token: String,
    val platform: String = "android",
)

/** Body for `POST /api/users/password`. The Joi schema at
 *  `backend/routes/users.js:736` uses camelCase, so the field names
 *  go on the wire unchanged. `currentPassword` is optional (omit for
 *  OAuth-only accounts setting an initial password). */
@JsonClass(generateAdapter = true)
data class PasswordUpdateBody(
    val currentPassword: String?,
    val newPassword: String,
)

/** Body for `POST /api/users/resend-verification`
 *  (`backend/routes/users.js:3049`). */
@JsonClass(generateAdapter = true)
data class ResendVerificationBody(
    val email: String,
)

/** Envelope from `GET /api/users/auth-methods`. */
@JsonClass(generateAdapter = true)
data class AuthMethodsResponse(
    val methods: List<AuthMethodDto>? = null,
    @Json(name = "has_password") val hasPassword: Boolean? = null,
    val providers: List<String>? = null,
    @Json(name = "two_factor_enabled") val twoFactorEnabled: Boolean? = null,
)

@JsonClass(generateAdapter = true)
data class AuthMethodDto(
    val id: String,
    val provider: String? = null,
    val label: String? = null,
)
